### PR TITLE
WIP: Live 1792: Synchronized onboarding logic

### PR DESCRIPTION
**Pull request by alexandremgo**
### ❓ Context

This PR is WIP (don't bother with all the `console.log`) but ready for comments and feedbacks👷

- **Impacted projects**: [LIVE-1792](https://ledgerhq.atlassian.net/browse/LIVE-1792?atlOrigin=eyJpIjoiYWFkNzgyMzI4YmM3NGYyZGI2YmYzNjlmNDg3NDlkYTciLCJwIjoiaiJ9)
- **Linked resource(s)**: 

The main features are:
- Function to extract the device onboarding state from byte flags
- Polling mechanism to retrieve the device onboarding state 
- Make the polling mechanism available as a react hook for LLM and LLD
- First implementation/POC of the synchronous onboarding in LLM:
  - BLE pairing of not seeded device
  - Screen to display the current onboarding state of the device using the above hook

🕵️‍♂️ Compared to the previous PR on the live-common and ledger-live-mobile, this feature fixed (parts, some are still WIP) [comments from this PR](https://github.com/LedgerHQ/ledger-live-mobile/pull/2510) and:
- In polling mechanism: replace `timer` that would push for a request every period time, without waiting for the previous request -> now custom `Observable` is created with a `repeat` 
  - the `new Observable` way has been chosen, compared to just used rxjs operators etc., with a system to **not delay** for the first run (and then repeat with a delay of X ms) because `repeat({ delay: (count) => count > 0 ? timer(periodMs) : timer(0) })` does not exist in rxjs v6
- Refactor the retry strategy: before on error was a "black box" where the system was not notified of an error, but was just retrying -> now a `allowedError` and a `fatalError` is returned from the polling mechanism and the hook so the app (LLM and LLD) are notified when an error occurred 

### ✅ Checklist

- [x] **Test coverage**: _Did you write any tests to cover the changes introduced by this pull request?_
  - [x]: Unit tests for the extract function & the hook polling function 
- [x] **Atomic delivery**: _Is this pull request standalone? In order words, does it depend on nothing else?_
- [ ] **No breaking changes**: _Does this pull request contain breaking changes of any kind? If so, please explain why._

### 📸 Demo

Will record a new demo video after on slack 

### 🚀 Expectations to reach

_Please make sure you follow these [**Important Steps**](https://github.com/LedgerHQ/ledger-live/blob/develop/CONTRIBUTING.md#important-steps)._

_Pull Requests must pass the CI and be internally validated in order to be merged._

### CODE REVIEW

File: apps/ledger-live-mobile/src/screens/SyncOnboarding/index.tsx [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        setStepIndex(5);
+        break;
+      case OnboardingStep.SafetyWarning:
+        setStepIndex(6);
+        break;
+      case OnboardingStep.Ready:
+        setStepIndex(7);
+        break;
+      default:
+        setStepIndex(0);
```
Comment (gre) [inline apps/ledger-live-mobile/src/screens/SyncOnboarding/index.tsx:120]: there may be a way to simplify this with an array containing all the steps that you are currently switch()ing on.

```js
const steps = [OnboardingStep.WelcomeScreen, ...]
const stepIndex = steps.indexOf(onboardingState?.currentOnboardingStep) + 1
```

**Review by gre — COMMENTED** [review]
really cool to see this happening 👏

File: apps/ledger-live-mobile/src/screens/SyncOnboarding/index.tsx (hunk 1 above)
Comment (alexandremgo) [reply apps/ledger-live-mobile/src/screens/SyncOnboarding/index.tsx:120]: Yeah, definitely I'll try to make something simpler like this ! Right now for this screen it was to demo and show that the polling mechanism was working ;)

File: libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+  // getDelayedOnboardingStateOnce
+  const getDelayedOnboardingStateOnce: Observable<OnboardingStatePollingResult> =
+    new Observable((subscriber) => {
+      console.log(`SyncOnboarding: ▶️ Polling from Observable ${i}`);
+      const delayMs = i > 0 ? pollingPeriodMs : 0;
+      console.log(`SyncOnboarding: polling delayed by ${delayMs} ms`);
+      i++;
+
+      const getOnboardingStateOnce =
+        (): Observable<OnboardingStatePollingResult> =>
```
Comment (grsoares21) [inline libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts:122]: Just a personal preference, but maybe letting typescript infer the type by itself will make the code cleaner.

File: libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts [hunk 3]
```diff
@@ -0,0 +1,246 @@
...
+          withDevice(deviceId)((t) => from(getVersion(t))).pipe(
+            // TODO: choose timeout ms value. For now = polling period
+            timeout(pollingPeriodMs), // Throws a TimeoutError
+            first(),
+            catchError((error: any) => {
+              if (isAllowedOnboardingStatePollingError(error)) {
+                // Pushes the error to the next step to be processed (no retry from the beginning)
```
Comment (grsoares21) [inline libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts:129]: You can maybe use partition and merge here. To have two different maps for handling errors and state.

File: libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+        });
+      }, delayMs);
+    });
+
+  return getDelayedOnboardingStateOnce.pipe(repeat());
+};
+
+// TODO: decide which errors are allowed
+export const isAllowedOnboardingStatePollingError = (
+  error: Error | any
```
Comment (grsoares21) [inline libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts:213]: Not sure why it's error | any here. But I'd say that this type of function should only accept Error, and if it's not an error it's up to the caller to check before passing it down to the function.

File: libs/ledger-live-common/src/hw/extractOnboardingState.ts [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+};
+
+export const extractOnboardingState = (flagsBytes: Buffer): OnboardingState => {
+  if (!flagsBytes || flagsBytes.length < onboardingFlagsBytesLength) {
+    throw new DeviceExtractOnboardingStateError(
+      "Incorrect onboarding flags bytes"
+    );
+  }
+
+  const isOnboarded = !!(flagsBytes[0] & onboardedMask);
```
Comment (grsoares21) [inline libs/ledger-live-common/src/hw/extractOnboardingState.ts:63]: you can also use Boolean(flagsBytes[0] & onboardedMask) here, I think. Will make it more explicit.

**Review by grsoares21 — COMMENTED** [review]
Looking good. Just a few comments on code clarity / style.

**Comment by alexandremgo** [comment]
FYI: i will move the `onboardingStatePolling` used by the hook to `live-common/src/hw/getOnboardingStatePolling` to keep the file containing the hook `useOnboardingStatePolling.ts` simple

File: libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts (hunk 4 above)
Comment (alexandremgo) [reply libs/ledger-live-common/src/onboarding/hooks/useOnboardingStatePolling.ts:236]: agreed, done ;)

File: apps/ledger-live-mobile/fastlane/Fastfile [hunk 6]
```diff
@@ -109,7 +109,7 @@ platform :ios do
 
     build_number = latest_testflight_build_number(
       version: trim_version_number(package["version"]),
-      app_identifier: MY_APP_BUNDLE_ID
+      app_identifier: "com.ledger.live"
```
Comment (juan-cortes) [inline apps/ledger-live-mobile/fastlane/Fastfile:112]: Why is this here?

File: apps/ledger-live-mobile/src/screens/PairDevices/index.js [hunk 7]
```diff
@@ -110,7 +116,26 @@ function PairDevicesInner({ navigation, route }: Props) {
...
+      if (route.params?.onlySelectDeviceWithoutFullAppPairing) {
+        // Sends a first request to trigger the native bluetooth pairing step
+        // (if the device is unknown to the phone) and make sure that the device
+        // is correctly paired to the phone
+        try {
+          await withDevice(device.deviceId)(t =>
+            from(getVersion(t)),
+          ).toPromise();
+        } catch (_) {
+          // Silently swwallowing error, we only want to trigger the native ble pairing step
```
Comment (juan-cortes) [inline apps/ledger-live-mobile/src/screens/PairDevices/index.js:132]: ww

File: apps/ledger-live-mobile/src/screens/PairDevices/index.js [hunk 8]
```diff
@@ -204,8 +229,17 @@ function PairDevicesInner({ navigation, route }: Props) {
 
   const onDone = useCallback(
     (device: Device) => {
-      navigation.goBack();
-      route.params?.onDone?.(device);
+      // To avoid passing a onDone function param that is not serializable
+      if (route.params?.onDoneNavigateTo === ScreenName.SyncOnboardingWelcome) {
+        console.log("PairDevices: 🦮 navigate directly to SyncOnboarding");
```
Comment (juan-cortes) [inline apps/ledger-live-mobile/src/screens/PairDevices/index.js:234]: 🪓

File: apps/ledger-live-mobile/src/screens/SyncOnboarding/index.tsx [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+  const { onboardingState, allowedError, fatalError } = useOnboardingStatePolling({ device, pollingPeriodMs });
+
+  const { pairedDevice } = route.params; 
+
+  // When reaching the synchronized onboarding, the device could already
+  // be BLE paired to the phone and at the same time not known by LLM,
+  // it does not matter.
+  // So as long as the device is not onboarded, the LLM will ask for a pairing.
+  useEffect(() => {
+    console.log("SyncOnboarding: navigate to pairDevices");
```
Comment (juan-cortes) [inline apps/ledger-live-mobile/src/screens/SyncOnboarding/index.tsx:39]: 🪓

File: apps/ledger-live-mobile/src/screens/SyncOnboarding/index.tsx (hunk 1 above)
Comment (juan-cortes) [reply apps/ledger-live-mobile/src/screens/SyncOnboarding/index.tsx:123]: +1

**Review by juan-cortes — COMMENTED** [review]
lgtm, so many tests, cleanup the logs and such though

File: apps/ledger-live-mobile/fastlane/Fastfile (hunk 6 above)
Comment (alexandremgo) [reply apps/ledger-live-mobile/fastlane/Fastfile:112]: it was a fixed to be able to build iOS that Valentin told me to try -> as this PR won't be merged, i did not put any comment on it. But Valentin merged this fix on develop yesterday ;)

**Pull request closed without merging**